# Project 12 -- Anthony Micco

**TA Help:** N/A

**Collaboration:** N/A

**Internet Resources:** None

**ChatGPT, Gemini, Claude, etc:** None

**Link to AI Chat History**: None

**OVERALL MESSAGE:** Any time that you used anything except your brain to solve the questions in these projects, you need to disclose such resources at the start of the project, with details about your usage of the tools.

**YOUR OWN WORK:** Even when you utilize other resources, do NOT just copy and paste.  Write all explanations in your own words, using several sentences in English, which are understandable and which you wrote (and did not just copy and paste).

## Question 1

In [1]:
# example of looping through arrays using the awk command

In [3]:
%%bash
echo "apple banana apple orange banana apple" | awk '
{
    # Loop through each word in the input
    for (i = 1; i <= NF; i++) {
        # Increment the count for the word
        # $i is the current word
        # count[$i] is the count for the word
        # ++ is the increment operator
        count[$i]++
    }
}
END {
    print "Word counts:"
    for (word in count) {
        print word ": " count[word]
    }
}'

Word counts:
banana: 2
orange: 1
apple: 3


In [4]:
# finding the customer with the most orders

In [6]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F',' '
NR > 1 {  # Skip header row
    customer_id = $2
    if (customer_id != "") {
        order_count[customer_id]++
    }
}
END {
    print "Customer ID, Number of Orders"
    for (customer in order_count) {
        print customer "," order_count[customer]
    }
}' | sort -t, -k2,2nr | head -n 10

XW90EAP,262
TL7Z2DM,151
VDEJEMP,146
HJFTTGW,128
B28LJKM,119
2JJAO46,115
KXXRD9J,114
TVSPG2H,113
BUIAFFT,109
1P5RQ9H,107


In [9]:
# looking at the structure of the orders.csv file

In [10]:
%%bash
head -n 3 /anvil/projects/tdm/data/restaurant/orders.csv

akeed_order_id,customer_id,item_count,grand_total,payment_mode,promo_code,vendor_discount_amount,promo_code_discount_percentage,is_favorite,is_rated,vendor_rating,driver_rating,deliverydistance,preparationtime,delivery_time,order_accepted_time,driver_accepted_time,ready_for_pickup_time,picked_up_time,delivered_time,delivery_date,vendor_id,created_at,LOCATION_NUMBER,LOCATION_TYPE,CID X LOC_NUM X VENDOR
163238.0,92PEE24,1.0,7.6,2,,0.0,,,No,,0.0,0.0,,,,,,,,2019-07-31 05:30:00,105,2019-08-01 05:30:16,0,,92PEE24 X 0 X 105
163240.0,QS68UD8,1.0,8.7,1,,0.0,,,No,,0.0,0.0,,,,,,,,2019-07-31 05:30:00,294,2019-08-01 05:31:10,0,Work,QS68UD8 X 0 X 294


In [7]:
# finding the vendor with the most orders

In [8]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F',' '
NR > 1 {  # Skip header row
    vendor_id = $22
    if (vendor_id != "") {
        order_count[vendor_id]++
    }
}
END {
    print "Vendor ID, Number of Orders"
    for (vendor in order_count) {
        print vendor "," order_count[vendor]
    }
}' | sort -t, -k2,2nr | head -n 10

113,7807
105,5562
79,5117
84,5001
78,4643
83,3684
386,3280
86,2837
846,2559
106,2263


1.2) The customer with the most orders was customer XW90EAP with 262 orders. \
1.3) The vendor with the most orders was vendor 113 with 7807 orders.

## Question 2

In [11]:
# checking to see the structure of the test_full.csv data set

In [13]:
%%bash
head -n 3 /anvil/projects/tdm/data/restaurant/test_full.csv

customer_id,gender,status_x,verified_x,created_at_x,updated_at_x,location_number,location_type,latitude_x,longitude_x,id,authentication_id,latitude_y,longitude_y,vendor_category_en,vendor_category_id,delivery_charge,serving_distance,is_open,OpeningTime,OpeningTime2,prepration_time,commission,is_akeed_delivering,discount_percentage,status_y,verified_y,rank,language,vendor_rating,sunday_from_time1,sunday_to_time1,sunday_from_time2,sunday_to_time2,monday_from_time1,monday_to_time1,monday_from_time2,monday_to_time2,tuesday_from_time1,tuesday_to_time1,tuesday_from_time2,tuesday_to_time2,wednesday_from_time1,wednesday_to_time1,wednesday_from_time2,wednesday_to_time2,thursday_from_time1,thursday_to_time1,thursday_from_time2,thursday_to_time2,friday_from_time1,friday_to_time1,friday_from_time2,friday_to_time2,saturday_from_time1,saturday_to_time1,saturday_from_time2,saturday_to_time2,primary_tags,open_close_flags,vendor_tag,vendor_tag_name,one_click_vendor,country_id,city_id,created_at_y,updat

In [14]:
#counting how many restaurants are in each category

In [15]:
%%bash
cat /anvil/projects/tdm/data/restaurant/test_full.csv | awk -F',' '
NR > 1 {  # Skip header row
    category = $15
    if (category != "") {
        category_count[category]++
    }
}
END {
    print "Category, Number of Restaurants"
    for (cat in category_count) {
        print cat "," category_count[cat]
    }
}' | sort -t, -k2,2nr | head -n 10

Restaurants,1435896
Sweets & Bakes,195804
Category, Number of Restaurants


In [21]:
# finding the unique category values in the test_full.csv

In [23]:
%%bash
awk -F',' '{print $15}' /anvil/projects/tdm/data/restaurant/test_full.csv | sort | uniq


Restaurants
Sweets & Bakes
vendor_category_en


In [16]:
#counting the number of restaurants with each rating

In [31]:
%%bash
cat /anvil/projects/tdm/data/restaurant/test_full.csv | awk -F',' '
NR > 1 {  # Skip header row
    rating = $30
    if (rating != "") {
        rating_count[rating]++
    }
}
END {
    print "Rating, Number of Restaurants"
    for (rate in rating_count) {
        print rate "," rating_count[rate]
    }
}' | sort -t, -k2,2nr 

4.5,310023
4.3,293706
4.2,244755
4.4,228438
4.6,163170
4.7,114219
4.0,97902
4.1,81585
3.8,32634
4.8,32634
3.2,16317
3.7,16317
Rating, Number of Restaurants


2.2) The category with the most restaurants is the restaurants category with 1435896 restaurants \
2.3) There are two different categories in total: restaurants and Sweets & Bakes. \
2.4) There are 97902 restaurants with a rating of 4.0, but there are no restaurants with a 5.0 rating.

## Question 3

In [24]:
# finding the average order amount by customer

In [26]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F',' '
NR > 1 {  # Skip header row
    customer_id = $2
    order_amount = $4 + 0  # Convert string to number by adding 0
    if (customer_id != "" && order_amount > 0) { # ensure that the customer and order actually exist
        total_spent[customer_id] += order_amount
        order_count[customer_id]++
    }
}
END {
    print "Customer ID, Total Spent, Number of Orders, Average Order"
    for (customer in total_spent) {
        avg = total_spent[customer] / order_count[customer]
        printf "%s,%.2f,%d,%.2f\n", customer, total_spent[customer], order_count[customer], avg
    }
}' | sort -t, -k4,4nr | head -n 10

HPERFQJ,783.80,1,783.80
AAJY8DZ,212.00,1,212.00
L4Y9P84,169.70,1,169.70
X6416DV,152.00,1,152.00
J2CLYW5,143.50,1,143.50
PUUEW3U,136.60,1,136.60
O1UCOD0,134.30,1,134.30
OT5DK5L,134.00,1,134.00
AN7TY2S,130.30,1,130.30
C31LYDS,130.30,1,130.30


In [27]:
#finding the customer that spent the most money

In [30]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F',' '
NR > 1 {  # Skip header row
    customer_id = $2
    order_amount = $4 + 0  # Convert string to number by adding 0
    if (customer_id != "" && order_amount > 0) { # ensure that the customer and order actually exist
        total_spent[customer_id] += order_amount
        order_count[customer_id]++
    }
}
END {
    print "Customer ID, Total Spent, Number of Orders"
    for (customer in total_spent) {
        printf "%s,%.2f,%d\n", customer, total_spent[customer], order_count[customer]
    }
}' | sort -t, -k2,4nr | head -n 10

B28LJKM,3965.50,119
XW90EAP,3830.60,261
UO5X10D,3274.60,98
C8D4BO3,2997.80,50
VEGWXS8,2558.80,72
BUIAFFT,2416.30,108
K3WSGUM,2390.60,72
ZCP73QX,2299.40,43
8771JHK,2231.50,100
2JJAO46,2215.80,115


In [32]:
#finding the average number of items per order from vendors

In [2]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F',' '
NR > 1 {  # Skip header row
    vendor_id = $22
    quantity = $3 + 0  # Convert string to number by adding 0
    if (vendor_id != "" && quantity > 0) { # ensure that the customer and order actually exist
        total_items[vendor_id] += quantity
        order_count[vendor_id]++
    }
}
END {
    print "Vendor ID, Total Items, Number of Orders, Average Items"
    for (vendor in total_items) {
        avg = total_items[vendor] / order_count[vendor]
        printf "%s,%d,%d,%.2f\n", vendor, total_items[vendor], order_count[vendor], avg
    }
}' | sort -t, -k4,4nr | head -n 10

180,2134,564,3.78
66,1915,526,3.64
159,6904,1970,3.50
33,5974,1822,3.28
43,3245,996,3.26
176,2458,758,3.24
274,1792,557,3.22
55,2306,725,3.18
67,1860,591,3.15
4,6106,1987,3.07


3.2) The customer with the highest average order amount if customer HPERFQJ with an average order of \\$783.80. \
3.3) The highest average order amount is \\$783.80. \
3.4) The customer that spent the most money was B28LJKM with \\$3965.50 \
3.5) The vendor with the highest average number of items per order is vendor 180 with 3.78 items per order.

## Question 4

In [3]:
# examining the structure of the 1995 flights dataset

In [5]:
%%bash
head -n 3 /anvil/projects/tdm/data/flights/subset/1995.csv

Year,Month,DayofMonth,DayOfWeek,DepTime,CRSDepTime,ArrTime,CRSArrTime,UniqueCarrier,FlightNum,TailNum,ActualElapsedTime,CRSElapsedTime,AirTime,ArrDelay,DepDelay,Origin,Dest,Distance,TaxiIn,TaxiOut,Cancelled,CancellationCode,Diverted,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay
1995,1,6,5,657,645,952,937,UA,482,N7298U,115,112,83,15,12,ORD,PHL,678,7,25,0,NA,0,NA,NA,NA,NA,NA
1995,1,7,6,648,645,938,937,UA,482,N7449U,110,112,88,1,3,ORD,PHL,678,5,17,0,NA,0,NA,NA,NA,NA,NA


In [6]:
#finding the average distance traveled each flight for each airline

In [10]:
%%bash
cat /anvil/projects/tdm/data/flights/subset/1995.csv | awk -F',' '
NR > 1 {
    airline_id = $9
    distance = $19 + 0
    if(airline_id != "" && distance > 0){
        total_distance[airline_id] += distance
        flight_count[airline_id]++
    }
}
END {
    print "Airline ID, Total Distance, # of Flights, Avg Distance"
    for(airline in total_distance){
        avg_distance = total_distance[airline] / flight_count[airline]
        printf "%s,%d,%d,%.2f\n", airline, total_distance[airline], flight_count[airline], avg_distance 
    }
    
}' | sort -t, -k4,4nr | head -n 10

AA,683619691,687987,993.65
UA,621849199,724441,858.38
CO,334020121,434425,768.88
TW,197919256,271766,728.27
AS,96104161,135124,711.23
NW,356173635,520727,683.99
HP,132070386,194508,679.00
DL,576699652,882351,653.59
US,426550655,778835,547.68
WN,275938737,691284,399.17


4.2) The carries with the highest average distance per flight is American Airlines with 993.65 miles per flight \
4.3) The highest average distance per flight is 993.65 miles

## Question 5

In [11]:
# checking the structure of the orders.csv file to see what columns I want to use for stats tracking

In [12]:
%%bash
head -n 3 /anvil/projects/tdm/data/restaurant/orders.csv

akeed_order_id,customer_id,item_count,grand_total,payment_mode,promo_code,vendor_discount_amount,promo_code_discount_percentage,is_favorite,is_rated,vendor_rating,driver_rating,deliverydistance,preparationtime,delivery_time,order_accepted_time,driver_accepted_time,ready_for_pickup_time,picked_up_time,delivered_time,delivery_date,vendor_id,created_at,LOCATION_NUMBER,LOCATION_TYPE,CID X LOC_NUM X VENDOR
163238.0,92PEE24,1.0,7.6,2,,0.0,,,No,,0.0,0.0,,,,,,,,2019-07-31 05:30:00,105,2019-08-01 05:30:16,0,,92PEE24 X 0 X 105
163240.0,QS68UD8,1.0,8.7,1,,0.0,,,No,,0.0,0.0,,,,,,,,2019-07-31 05:30:00,294,2019-08-01 05:31:10,0,Work,QS68UD8 X 0 X 294


In [17]:
#finding the average vendor discount amount for each vendor 

In [ ]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F',' '
NR > 1 {  # Skip header row
    vendor_id = $22
    discount_amt = $7+0
    if (vendor_id != "" && discount_amt > 0) {
        total_discount[vendor_id] += discount_amt
        order_count[vendor_id]++
    }
}
END {
    print "Vendor ID, Total Discount Amt, Order Count, Average Discount"
    for (vendor in order_count) {
        avg = total_discount[vendor] / order_count[vendor]
        printf "%s,%.2f,%d,%.2f\n", vendor, total_discount[vendor], order_count[vendor], avg
    }
}' | sort -t, -k4,4nr | head -n 10

In [1]:
#finding the highest average vendor rating

In [3]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F',' '
NR > 1 {  # Skip header row
    vendor_id = $22
    vendor_rating = $11+0
    if (vendor_id != "" && vendor_rating > 0) {
        total_rating[vendor_id] += vendor_rating
        order_count[vendor_id]++
    }
}
END {
    print "Vendor ID, Order Count, Average Rating"
    for (vendor in order_count) {
        avg = total_rating[vendor] / order_count[vendor]
        printf "%s,%d,%.2f\n", vendor,order_count[vendor], avg
    }
}' | sort -t, -k3,4nr | head -n 10

295,126,4.82
303,86,4.79
216,213,4.76
115,48,4.75
13,114,4.75
294,19,4.74
113,1030,4.71
265,20,4.70
310,220,4.70
79,1108,4.68


5.2) For my first example, I found the vendors that offered the highest average discount amount. For this, I used the vendor id column (\\$22) and the vendor_discount_amount column (\\$7) to find the total discount amount and the total number of orders for each vendor. I then iterated the total number of orders and found the average discount amount for the vendors. From this, I found that vendor 85 had the highest average discount amount of $3.33. This information may be useful because it allows customers to find vendors that offer higher discounts so they do not have to pay as much for their orders. Then, for my second example, I found the vendors with the highest average rating. To do this, I again used the vendor id column, but this time, I also used the vendor_rating column (\\$11) to find the total rating and the total number of orders for each vendor. I again iterated through the total orders and found the average rating for the vendors, where I found that vendor 295 had the highest average rating of 4.82. This information can be useful because customers will want to know which vendor is the best to order from when placing an order. \
5.3) I used associative arrays to accomplish this because I used the vendor_id column as the indices for the counts, as I wanted to calculate statistics by grouping each unique vendor together based on their id. 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
